// nickjs/lib/es5/chrome/BrowserDriver.js
"use strict";var _keys = require("babel-runtime/core-js/object/keys");var _keys2 = _interopRequireDefault(_keys);var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);var _createClass2 = require("babel-runtime/helpers/createClass");var _createClass3 = _interopRequireDefault(_createClass2);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}



var TabDriver = require("./TabDriver");

var CDP = require('chrome-remote-interface');
var _ = require("lodash");var

BrowserDriver = function () {

	function BrowserDriver(nick) {(0, _classCallCheck3.default)(this, BrowserDriver);
		this._nick = nick;
		this._options = nick.options;
	}(0, _createClass3.default)(BrowserDriver, [{ key: "exit", value: function exit(

		code) {
			process.exit(code);
		} }, { key: "_initialize", value: function _initialize(

		callback) {

























			var execFile = require("child_process").execFile;
			var chromePath = process.env.CHROME_PATH || "google-chrome-unstable";
			var child = execFile(chromePath, [
			"--remote-debugging-port=9222",

			"--disable-gpu",
			"--headless",
			"--hide-scrollbars",

			"--disable-web-security",
			"--allow-insecure-localhost",
			"--allow-running-insecure-content",
			"--allow-file-access-from-files", "--window-size=" +

			this._options.width + "," + this._options.height,

			"--disable-translate",
			"--disable-extensions",
			"--disable-sync",
			"--disable-background-networking",
			"--safebrowsing-disable-auto-update",
			"--metrics-recording-only",
			"--disable-default-apps",
			"--no-first-run",
			"--no-sandbox"]);

			process.on("exit", function () {
				try {
					child.kill();
				} catch (e) {}
			});
			child.on("error", function (err) {
				callback("could not start chrome: " + err);
			});
			var cleanSocket = function cleanSocket(socket) {
				socket.removeAllListeners();
				socket.end();
				socket.destroy();
				socket.unref();
			};
			var net = require("net");
			var checkStart = Date.now();
			var nbChecks = 0;
			var checkDebuggerPort = function checkDebuggerPort() {
				setTimeout(function () {
					var socket = net.createConnection(9222);
					socket.once("error", function (err) {
						++nbChecks;
						cleanSocket(socket);
						if (Date.now() - checkStart > 5 * 1000) {
							callback("could not connect to chrome debugger after " + nbChecks + " tries: " + err);
						} else {
							checkDebuggerPort();
						}
					});
					socket.once("connect", function () {
						cleanSocket(socket);
						callback(null);
					});
				}, 200);
			};
			checkDebuggerPort();
		} }, { key: "_newTabDriver", value: function _newTabDriver(

		uniqueTabId, callback) {var _this = this;
			var connectToTab = function connectToTab(cdpTarget) {
				CDP({ target: cdpTarget }, function (client) {
					var tab = new TabDriver(uniqueTabId, _this._options, client, cdpTarget.id);
					tab._init(function (err) {
						if (err) {
							callback(err);
						} else {
							callback(null, tab);
						}
					});
				}).on('error', function (err) {
					callback("cannot connect to chrome tab: " + err);
				});
			};
			CDP.List(function (err, tabs) {
				if (err) {
					callback("could not list chrome tabs: " + err);
				} else {
					if (uniqueTabId === 1 && tabs.length === 1 && tabs[0].url === "about:blank") {

						connectToTab(tabs[0]);
					} else {
						CDP.New(function (err, tab) {
							if (err) {
								callback("cannot create new chrome tab: " + err);
							} else {

								connectToTab(tab);
							}
						});
					}
				}
			});
		} }, { key: "__getOneTabForCookieRequest", value: function __getOneTabForCookieRequest(





		methodName, callback) {
			var tab = this._nick.tabs[(0, _keys2.default)(this._nick.tabs)[0]];
			if (tab) {
				return tab;
			} else {
				callback(methodName + ": could not manipulate cookies because there are no open tabs, please open at least one tab");
				return null;
			}
		} }, { key: "_getAllCookies", value: function _getAllCookies(

		callback) {
			var tab = this.__getOneTabForCookieRequest("getAllCookies", callback);
			if (tab) {
				tab.driver._getAllCookies(callback);
			}
		} }, { key: "_deleteAllCookies", value: function _deleteAllCookies(

		callback) {
			var tab = this.__getOneTabForCookieRequest("deleteAllCookies", callback);
			if (tab) {
				tab.driver._deleteAllCookies(callback);
			}
		} }, { key: "_deleteCookie", value: function _deleteCookie(

		name, domain, callback) {
			var tab = this.__getOneTabForCookieRequest("deleteCookie", callback);
			if (tab) {
				tab.driver._deleteCookie(name, domain, callback);
			}
		} }, { key: "_setCookie", value: function _setCookie(

		cookie, callback) {
			var tab = this.__getOneTabForCookieRequest("setCookie", callback);
			if (tab) {
				tab.driver._setCookie(cookie, callback);
			}
		} }]);return BrowserDriver;}();



module.exports = BrowserDriver;
